### PR TITLE
Added two-way ssl authentication

### PR DESCRIPTION
**Pull request by Oofnik**
Added two-way authentication capabilities on the SSL security layer on ESP32. 
Before this change only a call to setCACert() is available, which only allows one-way server authentication.
Two-way authentication can be done by allowing the usage of setCertificate() and setPrivateKey() in addition to setCACert() in the WiFiClientSecure class.

### CODE REVIEW

**Comment by gilmaimon** [comment]
Hi, I'm sorry it took me so long to get to this PR.
First, thank you for the contribution and pull request, I really appreciate it.

The thing which bothers me in this PR is that it is esp32 only, users with esp8266 will have problems trying to use this code. As far as I understand, because of compiler optimizations the code is able to compile as long as you don't need (use) the functions you added, but it does not meet the requirements of the library (which is full esp8266 and esp32 compatibility).

Bottom line is, I can't merge it right now because it will break the requirements for the library. 
I will keep the PR open in case anyone else will contribute the rest of the needed code.

Thanks,
Gil.

**Comment by Oofnik** [comment]
So I have added support to ESP8266 two-way authentication. Note that the function calls for setting the certificates/key are different than the ESP32.
Added an example file only to test compilation. Since I don't have an ESP8266 device I didn't actually test the code on an actual device.
Also note that we don't use the setFingerprint() in the example but we use the new functions that actually check the certificates (up to some BearSSL library limitations).

**Comment by gilmaimon** [comment]
Looks good, I'll do my best to test and merge it this weekend.

**Comment by developerfromjokela** [comment]
So is it working?

**Comment by developerfromjokela** [comment]
@gilmaimon?

**Comment by Oofnik** [comment]
@developerfromjokela 
I was testing it with ESP32. If you have ESP8266, why don't you test it?

**Comment by developerfromjokela** [comment]
I have it. I'll test it tomorrow.

**Comment by gilmaimon** [comment]
Sorry for not approving this. I really can't find the time to test it myself, but I also don't want it to stay still after you invested time and effort to make this work. This might not be very responsible, but I will merge this based only on the automatic tests and the code changes.
It seems like you did a great job following the library current codebase guidelines and I thank you very much for the contribuation.
Again I am sorry this took so long, thanks alot!

**Comment by adam-kral** [comment]
Hi, I tried to use setTrustAnchors, as in example, with echo.websocket.org. But my callbacks don't get called.

I copied the entire certificate chain I got from firefox when visiting [https://echo.websocket.org](https://echo.websocket.org). (Because I didn't know -- would just the websocket's cert work, or does it have to be the CA that signed it? Could you please answer this, too?) 

My exact code follows. Everything prints out except the callbacks.

```

#include <ArduinoWebsockets.h>
#include <ESP8266WiFi.h>

const char* ssid = "myssid"; //Enter SSID
const char* password = "mypass"; //Enter Password

const char* websockets_connection_string = "wss://echo.websocket.org/"; //Enter server adress

const char ca_cert[] PROGMEM = R"EOF(
-----BEGIN CERTIFICATE-----
MIIFZDCCBEygAwIBAgISA6/6YxIE8i7BOXUrWoRypU4XMA0GCSqGSIb3DQEBCwUAMEoxCzAJBgNVBAYTAlVTMRYwFAYDVQQKEw1MZXQncyBFbmNyeXB0MSMwIQYDVQQDExpMZXQncyBFbmNyeXB0IEF1dGhvcml0eSBYMzAeFw0yMDA1MjYyMTE5MDVaFw0yMDA4MjQyMTE5MDVaMBgxFjAUBgNVBAMTDXdlYnNvY2tldC5vcmcwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDZd7d/TtDlhUK380nCBsNicWyrEm8NF8gPjoDGlHxP5uOCwGC/xK1hL8oW8DpOTFVQEmyCQUlnysf1BX1esjkntE3W/mmaH4ybSz5j6tdSdKGcX1+qgyHPZt4B7HEHzBAKfk7FX0+63MZtR7fy/FtoB81UpHZgMoFG+2SRs/J0Z4doqWyGulEwO2GnwrM/AbyU4m4DMOzU+QPVJ6CpR/SyOzsyoyZBTv9xrIgbz3ML7Hyjo16o813rx7VCwIrnK0GI+QwrqoOY9IXBDjFxMJmI7CNo1lleXgwI+vAO/IOeWyoYOfHwmfWSbazUGOSBSUimormK5Kp6lXvt5JvcBjD/AgMBAAGjggJ0MIICcDAOBgNVHQ8BAf8EBAMCBaAwHQYDVR0lBBYwFAYIKwYBBQUHAwEGCCsGAQUFBwMCMAwGA1UdEwEB/wQCMAAwHQYDVR0OBBYEFMvYlDHLSsJjYEIhYU/CplNRszxuMB8GA1UdIwQYMBaAFKhKamMEfd265tE5t6ZFZe/zqOyhMG8GCCsGAQUFBwEBBGMwYTAuBggrBgEFBQcwAYYiaHR0cDovL29jc3AuaW50LXgzLmxldHNlbmNyeXB0Lm9yZzAvBggrBgEFBQcwAoYjaHR0cDovL2NlcnQuaW50LXgzLmxldHNlbmNyeXB0Lm9yZy8wKQYDVR0RBCIwIIIPKi53ZWJzb2NrZXQub3Jngg13ZWJzb2NrZXQub3JnMEwGA1UdIARFMEMwCAYGZ4EMAQIBMDcGCysGAQQBgt8TAQEBMCgwJgYIKwYBBQUHAgEWGmh0dHA6Ly9jcHMubGV0c2VuY3J5cHQub3JnMIIBBQYKKwYBBAHWeQIEAgSB9gSB8wDxAHYAXqdz+d9WwOe1Nkh90EngMnqRmgyEoRIShBh1loFxRVgAAAFyUxFZ+gAABAMARzBFAiEA8QDXE/90IRkbZYovyCDdla91fbzwIe4ZYAgOJtYanvECIE2QMz3cOTsTtCw+NYvRyew4wi1eHLEYJjoq/oZZca0uAHcAsh4FzIuizYogTodm+Su5iiUgZ2va+nDnsklTLe+LkF4AAAFyUxFaBAAABAMASDBGAiEAsFtUItanUOdR4INS+ogNXOfaH3f7C4ff0MG+9nopgIwCIQC0jawZ1VCs5GuYHBhoeWr/vQFSc3hzQD57jZR9XJXdejANBgkqhkiG9w0BAQsFAAOCAQEAYx5L1OI9/5wKB2sW7weuT/xDcJTOdaawCnMzUJgF9RRJrN5bacdbdt5wKsajn6Q1Q7XalhzuaIuPOlMEgOpn26aoQgksfiz21s+ZEzEK/HGG+BuHoNrGd2iI+dmFFWyDQrI+ILnubgNelDvFahabTrnrsECy+D5lz1byPHxUgMkzWP0oS0+BoXPYq8RmnXmLGwsRha2FbpSXUNjqO8IbHqyRi/e/aW1gdjwMPVZ702uhIrYDhHekm7zH1HwMTD6FwRdok5bec/C81dD+LrFCY8qRAWcOl7gaC6O+s4LmpaF6ViBX4iO9KhXy6Y+qC1b0UMW5zt6n0jggJo5yYfCmew==
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
MIIEkjCCA3qgAwIBAgIQCgFBQgAAAVOFc2oLheynCDANBgkqhkiG9w0BAQsFADA/MSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMTDkRTVCBSb290IENBIFgzMB4XDTE2MDMxNzE2NDA0NloXDTIxMDMxNzE2NDA0NlowSjELMAkGA1UEBhMCVVMxFjAUBgNVBAoTDUxldCdzIEVuY3J5cHQxIzAhBgNVBAMTGkxldCdzIEVuY3J5cHQgQXV0aG9yaXR5IFgzMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAnNMM8FrlLke3cl03g7NoYzDq1zUmGSXhvb418XCSL7e4S0EFq6meNQhY7LEqxGiHC6PjdeTm86dicbp5gWAf15Gan/PQeGdxyGkOlZHP/uaZ6WA8SMx+yk13EiSdRxta67nsHjcAHJyse6cF6s5K671B5TaYucv9bTyWaN8jKkKQDIZ0Z8h/pZq4UmEUEz9l6YKHy9v6Dlb2honzhT+Xhq+w3Brvaw2VFn3EK6BlspkENnWAa6xK8xuQSXgvopZPKiAlKQTGdMDQMc2PMTiVFrqoM7hD8bEfwzB/onkxEz0tNvjj/PIzark5McWvxI0NHWQWM6r6hCm21AvA2H3DkwIDAQABo4IBfTCCAXkwEgYDVR0TAQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8EBAMCAYYwfwYIKwYBBQUHAQEEczBxMDIGCCsGAQUFBzABhiZodHRwOi8vaXNyZy50cnVzdGlkLm9jc3AuaWRlbnRydXN0LmNvbTA7BggrBgEFBQcwAoYvaHR0cDovL2FwcHMuaWRlbnRydXN0LmNvbS9yb290cy9kc3Ryb290Y2F4My5wN2MwHwYDVR0jBBgwFoAUxKexpHsscfrb4UuQdf/EFWCFiRAwVAYDVR0gBE0wSzAIBgZngQwBAgEwPwYLKwYBBAGC3xMBAQEwMDAuBggrBgEFBQcCARYiaHR0cDovL2Nwcy5yb290LXgxLmxldHNlbmNyeXB0Lm9yZzA8BgNVHR8ENTAzMDGgL6AthitodHRwOi8vY3JsLmlkZW50cnVzdC5jb20vRFNUUk9PVENBWDNDUkwuY3JsMB0GA1UdDgQWBBSoSmpjBH3duubRObemRWXv86jsoTANBgkqhkiG9w0BAQsFAAOCAQEA3TPXEfNjWDjdGBX7CVW+dla5cEilaUcne8IkCJLxWh9KEik3JHRRHGJouM2VcGfl96S8TihRzZvoroed6ti6WqEBmtzw3Wodatg+VyOeph4EYpr/1wXKtx8/wApIvJSwtmVi4MFU5aMqrSDE6ea73Mj2tcMyo5jMd6jmeWUHK8so/joWUoHOUgwuX4Po1QYz+3dszkDqMp4fklxBwXRsW10KXzPMTZ+sOPAveyxindmjkW8lGy+QsRlGPfZ+G6Z6h7mjem0Y+iWlkYcV4PIWL1iwBi8saCbGS5jN2p8M+X+Q7UNKEkROb3N6KOqkqm57TH2H3eDJAkSnh6/DNFu0Qg==
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
MIIDSjCCAjKgAwIBAgIQRK+wgNajJ7qJMDmGLvhAazANBgkqhkiG9w0BAQUFADA/MSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMTDkRTVCBSb290IENBIFgzMB4XDTAwMDkzMDIxMTIxOVoXDTIxMDkzMDE0MDExNVowPzEkMCIGA1UEChMbRGlnaXRhbCBTaWduYXR1cmUgVHJ1c3QgQ28uMRcwFQYDVQQDEw5EU1QgUm9vdCBDQSBYMzCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAN+v6ZdQCINXtMxiZfaQguzH0yxrMMpb7NnDfcdAwRgUi+DoM3ZJKuM/IUmTrE4Orz5Iy2Xu/NMhD2XSKtkyj4zl93ewEnu1lcCJo6m67XMuegwGMoOifooUMM0RoOEqOLl5CjH9UL2AZd+3UWODyOKIYepLYYHsUmu5ouJLGiifSKOeDNoJjj4XLh7dIN9bxiqKqy69cK3FCxolkHRyxXtqqzTWMIn/5WgTe1QLyNau7Fqckh49ZLOMxt+/yUFw7BZy1SbsOFU5Q9D8/RhcQPGX69Wam40dutolucbY38EVAjqr2m7xPi71XAicPNaDaeQQmxkqtilX4+U9m5/wAl0CAwEAAaNCMEAwDwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMCAQYwHQYDVR0OBBYEFMSnsaR7LHH62+FLkHX/xBVghYkQMA0GCSqGSIb3DQEBBQUAA4IBAQCjGiybFwBcqR7uKGY3Or+Dxz9LwwmglSBd49lZRNI+DT69ikugdB/OEIKcdBodfpga3csTS7MgROSR6cz8faXbauX+5v3gTt23ADq1cEmv8uXrAvHRAosZy5Q6XkjEGB5YGV8eAlrwDPGxrancWYaLbumR9YbK+rlmM6pZW87ipxZzR8srzJmwN0jP41ZL9c8PDHIyh8bwRLtTcm1D9SZImlJnt1ir/md2cXjbDaJWFBM5JDGFoqgCWjBH4d1QB7wCCZAA62RjYJsWvIjJEubSfZGL+T0yjWW06XyxV3bqxbYoOb8VZRzI9neWagqNdwvYkQsEjgfbKbYK7p2CNTUQ
-----END CERTIFICATE-----
)EOF";

using namespace websockets;

void onMessageCallback(WebsocketsMessage message) {
    Serial.print("Got Message: ");
    Serial.println(message.data());
}

void onEventsCallback(WebsocketsEvent event, String data) {
    if(event == WebsocketsEvent::ConnectionOpened) {
        Serial.println("Connnection Opened");
    } else if(event == WebsocketsEvent::ConnectionClosed) {
        Serial.println("Connnection Closed");
    } else if(event == WebsocketsEvent::GotPing) {
        Serial.println("Got a Ping!");
    } else if(event == WebsocketsEvent::GotPong) {
        Serial.println("Got a Pong!");
    }
}
    
X509List cert(ca_cert);


WebsocketsClient client;
void setup() {
    Serial.begin(115200);
    // Connect to wifi
    WiFi.begin(ssid, password);

    // Wait some time to connect to wifi
    for(int i = 0; i < 10 && WiFi.status() != WL_CONNECTED; i++) {
        Serial.print(".");
        delay(1000);
    }

    // run callback when messages are received
    client.onMessage(onMessageCallback);
    
    // run callback when events are occuring
    client.onEvent(onEventsCallback);

    client.setTrustAnchors(&cert);

    // Before connecting, set the ssl fingerprint of the server
//    client.setFingerprint(echo_org_ssl_fingerprint);

    // Connect to server
    client.connect(websockets_connection_string);
    Serial.println(client.getCloseReason());  // I get -1, so this works?

    // Send a message
    Serial.println(client.send("Hello Server"));
    Serial.println("after send");


    // Send a ping
    client.ping();
    Serial.println("after ping");

}

void loop() {
    client.poll();
}

```

**Comment by adam-kral** [comment]
I also tried it with an up-to-date SHA-1 fingerprint, `setFingerprint`, (not secure though) and that worked.
